### ISSUE TITLE
LIXA EOC's not working properly

### ISSUE THREAD

**Issue by CoroNaut** [CleverRaven/Cataclysm-DDA#76479]
### Describe the bug

(SPOILERS)
(SPOILERS)

I enter the LIXA "wall of flat space" but get unfolded effect and the popup, then nothing happens. I read some of the code, is it supposed to teleport the player? This doesn't seem intended that it just gives you the unfolded effect and never goes away.

I ended up walking around outside the facility and fought a few zombies to see if anything would happen, but no. I also tried deconstructing some parts of the machine, nothing. 

### Attach save file

[Saegertown-trimmed.tar.gz](https://github.com/user-attachments/files/17012457/Saegertown-trimmed.tar.gz)

### Steps to reproduce

1. Load save
2. Walk into "wall of flat space"
3. It gives "unfolded" effect, and pops up its cool message, but doesn't do anything else
4. Walk around in the real world fighting zombies outside the facility as if you were an invader into this dimension

### Expected behavior

It's supposed to work as intended (whatever intended is)

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19045.4651 (22H2)
- Game Version: cdda-experimental-2024-09-16-0441 e186f1b [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

**Comment by PGR-14** [comment]
Can confirm, doesn't always work as intended.

**Comment by IdleSol** [comment]
If I understand correctly. The problem is that there are no coordinates for teleportation.

1. based on this:
https://github.com/CleverRaven/Cataclysm-DDA/blob/182ec6c2943b5b9b3a723d4e691688cc8d3433c5/data/json/mapgen/LIXA_mapgen.json#L1233

This assumes that the `LIXA_forward_tp_1` value is assigned coordinates. Similarly for `LIXA_forward_tp_2` ... `LIXA_forward_tp_16`. And so 7 times for each parameter. I don't know if this is correct or not.

2. When the trap is triggered, the following should happen:
https://github.com/CleverRaven/Cataclysm-DDA/blob/182ec6c2943b5b9b3a723d4e691688cc8d3433c5/data/json/effects_on_condition/nether_eocs/LIXA_EOCs_spells_traps.json#L612

If I understand it correctly, it means teleporting the character by the coordinates of the **global variable** `LIXA_forward_tp_1`.

But there is a problem. If I display the list of global variables, there is no such variable in it. 

For the control check, open the debug menu and set the value of the global variable:
- LIXA_forward_tp_1
- 0,0,0

After that we enter the corresponding trap. The trap is triggered and the character is teleported to the specified coordinates.

In the save, trap 1 is at the top. And trap 16, respectively, is at the bottom.

**Comment by molkemon** [comment]
This is still in issue in experimental from today:

- OS: Windows
    - OS Version: 10.0.19045.5131 (22H2)
- Game Version: df0d134 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth],
    Bombastic Perks [bombastic_perks],
    SpeedyDex [speedydex],
    Stats Through Kills [stats_through_kills],
    Martial Mastery [perk_melee_system]
]


[Badlands-trimmed.tar.gz](https://github.com/user-attachments/files/18053669/Badlands-trimmed.tar.gz)


I have already walked into every one of the portals, every single one gave the same text pop up the first time, but no teleportation. I am now stuck unfolded and also can't escape the LIXA facility as I can't get out via the elevators. RIP

**Comment by molkemon** [comment]
Ok, after fiddling around with this for hours, I have come up with a workaround fix. I have changed the EOC_LIXA_UNFOLD_ENTRY EOC in such a way, that it generates the missing tp points, which will allow progression on running saves (something that the actual bugfix probably won't). 

This is not a solution for the actual bug. The actual bug is that worldgen should set these tp points as global vars, but for unknown reasons doesn't. After reading up on documentation, I'm pretty sure it should work as implemented, but oh well...

Anyways, here is the full modified EOC file, just drop it in data/json/effects_on_condition/nether_eocs/ and overwrite (or if you want to be save and run across this further in the future, just copy Lines 204 to 426 into the EOC_LIXA_UNFOLD_ENTRY effect. 

[LIXA_EOCs_spells_traps.json](https://github.com/user-attachments/files/18055376/LIXA_EOCs_spells_traps.json)


Take note: 
This file will specifically work for LIXA experiment chambers, where you enter the portals from east to west. It will probably not work out of the box for other orientations (in fact it will probably crash the game after you run around for a while inside the folded space). 

If you have a different orientation, you need to change a bit of stuff, starting on Line 204 until Line 426. You will see a u_location_variable effect for each of the 16 forward and backward traps/tps, so 32 blocks in total. A block looks like this:

```
      {
        "u_location_variable": { "global_val": "LIXA_forward_tp_1" }, "trap": "tr_LIXA_back_1", "target_max_radius": 400,
        "target_params": { "om_terrain": "LIXA_device_unfolded_A", "search_range": 400, "min_distance": 0, "z": 0 },
		"x_adjust": -1,
        "z_adjust": -8,
        "z_override": true
      },    
```

If you enter your portals from west to east, you have to change x_adjust to 1 (and the ones that are 1 already to -1)
If you enter north to south (not sure if this is even possible) then you have to change x_adjust 1 to y_adjust 1 (and the x -1 to y -1) and if you enter south to north then x_adjust 1 becomes y_adjust -1 (and x_adjust -1 becomes y_adjust 1).

You have to change this for all 32 blocks.

**Comment by Procyonae** [comment]
> This is not a solution for the actual bug. The actual bug is that mapgen should set these tp points as global vars, but for unknown reasons doesn't. After reading up on documentation, I'm pretty sure it should work as implemented, but oh well...

It is working as is at least sometimes, I've tried it six times so far and it's worked every time

**Comment by molkemon** [comment]
> It is working as is at least sometimes, I've tried it six times so far and it's worked every time

I don't know if this would make a difference, but in my world LIXA is pretty close to my initial spawn ie might have been genned during initial worldgen rather than later on through normal generation. Maybe it's also based on the orientation it gets genned in, I haven't looked at the rotation stuff. In my save Lixa main entrance goes from west to east and the experiment chamber is east to west (so you enter the portals from east to west). 

I have also used a text in file search to search for other mapgens that use the variable thingy to appoint global vars, and there is only one other thing in mapgen that does that, the creepy black glass farm thingy (vitrified farm or something). I also have this in my world. The vitrified thingy also sets a few points, and those were correctly set as global vars. The farm is a bit futher away from my spawn though so I'm pretty sure it was genned later, not during init.

**Comment by Regilith** [comment]
I also encountered this issue with my game. My LIXA lab is also nearby my spawn, so it's probably the same issue.

I tried molkemon's workaround, and I had to make a slight edit to make it work on the latest experimental: remove the `"search_range": 400,` from each block, so that each block looks like this:


```
      {
        "u_location_variable": { "global_val": "LIXA_forward_tp_1" }, "trap": "tr_LIXA_back_1", "target_max_radius": 400,
        "target_params": { "om_terrain": "LIXA_device_unfolded_A", "min_distance": 0, "z": 0 },
		"x_adjust": -1,
        "z_adjust": -8,
        "z_override": true
      },    
```


I also had to adjust my `x_adjust` to make it work with my west-east orientation.

I did not have to adjust line 427. Thank you @molkemon for the workaround!

**Comment by molkemon** [comment]
Yes, the quite farm that teleports you to the black/white obsidian glass dimension.

**Comment by gynther** [comment]
Can verify that there is still a problem, but instead of above problems it just teleported me to 0,0,0 on the overmap which incidentally was a sea. good thing i savescum i suppose...

Anyhow using the above instructions i fixed the problems and here is a current [LIXA_EOCs_spells_traps.json](https://github.com/user-attachments/files/20099912/LIXA_EOCs_spells_traps.json) that seems to work.

**Comment by epenalver** [comment]
> Can verify that there is still a problem, but instead of above problems it just teleported me to 0,0,0 on the overmap which incidentally was a sea. good thing i savescum i suppose...
> 
> Anyhow using the above instructions i fixed the problems and here is a current [LIXA_EOCs_spells_traps.json](https://github.com/user-attachments/files/20099912/LIXA_EOCs_spells_traps.json) that seems to work.

I encountered the same issue in experimental build 2025-06-09-1824 and the json attached here worked for me.
